src/protego.py

```diff
@@ -0,0 +1,337 @@
+            else:
+                # We will be generous here and give it a second chance.
+                parts = line.rsplit(' ', 1)
+                if not len(parts) == 2:
+                    continue
+                field, value = parts
+
+            field = field.strip().lower()
+            value = value.strip()
+
+            # Ignore rules with no value part (e.g. "Disallow: ", "Allow: ").
+            if not value:
+                previous_rule = Rule(field, value)
+                continue
+
+            # Ignore rules without a corresponding user agent.
+            if not current_user_agents and field not in USER_AGENT_DIRECTIVE:
+                continue
+
+            self._total_directive_seen += 1
+
+            if field in USER_AGENT_DIRECTIVE:
+                if previous_rule and previous_rule.field not in USER_AGENT_DIRECTIVE:
+                    current_user_agents = []
+
+                # Wildcards are not supported in the user agent values.
+                # We will be generous here and remove all the wildcards.
+                user_agent = value.strip().lower()
+                if user_agent != '*' and '*' in user_agent:
+                    user_agent = user_agent.replace('*', '')
+
+                # See if this user agent is encountered before, if so merge these rules into it.
+                rule_set = self._user_agents.get(user_agent, None)
+                if rule_set and rule_set not in current_user_agents:
+                    current_user_agents.append(rule_set)
+
+                if not rule_set:
+                    rule_set = _RuleSet()
+                    rule_set.user_agent = user_agent
+                    self._user_agents[user_agent] = rule_set
+                    current_user_agents.append(rule_set)
+
+            elif field in ALLOW_DIRECTIVE:
+                for user_agent in current_user_agents:
+                    user_agent.allow(value)
+
+            elif field in DISALLOW_DIRECTIVE:
+                for user_agent in current_user_agents:
+                    user_agent.disallow(value)
+
+            elif field in SITEMAP_DIRECTIVE:
+                self._sitemap_list.append(value)
+
+            elif field in CRAWL_DELAY_DIRECTIVE:
+                for user_agent in current_user_agents:
+                    user_agent.crawl_delay = value
+
+            elif field in REQUEST_RATE_DIRECTIVE:
+                for user_agent in current_user_agents:
+                    user_agent.request_rate = value
+
+            elif field in HOST_DIRECTIVE:
+                self._host = value
+
+            else:
+                self._invalid_directive_seen += 1
+
+            previous_rule = Rule(field, value)
+
+    def _get_matching_rule_set(self, user_agent):
+        """Return the rule set with highest matching score."""
+        if not self._user_agents:
+            return None
+        score_rule_set_pairs = ((rs.applies_to(user_agent), rs) for rs in self._user_agents.values())
```

Comment:
Maybe this should be cached for each new user_agent received, so that the rule set for a given user_agent only needs to be determined once per user_agent at run time.